File: src/java/classes/samples/AnalyticalReplicate.java
/* ***************************************************************
 *  This file is part of STATegra EMS.
 *
 *  STATegra EMS is free software: you can redistribute it and/or 
 *  modify it under the terms of the GNU General Public License as
 *  published by the Free Software Foundation, either version 3 of 
 *  the License, or (at your option) any later version.
 * 
 *  STATegra EMS is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 * 
 *  You should have received a copy of the GNU General Public License
 *  along with STATegra EMS.  If not, see <http://www.gnu.org/licenses/>.
 * 
 *  More info http://bioinfo.cipf.es/stategraems
 *  Technical contact stategraemsdev@gmail.com
 *  *************************************************************** */
package classes.samples;

import com.google.gson.Gson;

/**
 *
 * @author Rafa Hernández de Diego
 */
public class AnalyticalReplicate {

    private String analytical_rep_id;
    private String analytical_rep_name;
    private String bioreplicate_id;
    private String treatment_id;
    String status = null;

    /**
     *
     */
    public AnalyticalReplicate() {
    }

    /**
     *
     * @param analytical_rep_id
     * @param analytical_rep_name
     * @param bioreplicate_id
     * @param treatment_id
     */
    public AnalyticalReplicate(String analytical_rep_id, String analytical_rep_name, String bioreplicate_id, String treatment_id) {
        this.analytical_rep_id = analytical_rep_id;
        this.analytical_rep_name = analytical_rep_name;
        this.bioreplicate_id = bioreplicate_id;
        this.treatment_id = treatment_id;
    }

    /**
     * This static function returns a new AnalyticalReplicate object using the
     * data contained in the given JSON object (as String).
     *
     * @param jsonString the JSON object
     * @return the new Object.
     */
    public static AnalyticalReplicate fromJSON(String jsonString) {
        Gson gson = new Gson();
        AnalyticalReplicate analyticalReplicate = gson.fromJson(jsonString, AnalyticalReplicate.class);
        return analyticalReplicate;
    }

    /**
     * This function returns the object as a JSON format string.
     *
     * @return the object as JSON String
     */
    public String toJSON() {
        Gson gson = new Gson();
        String jsonString = gson.toJson(this);

        return jsonString;
    }

    //**********************************************************************
    //* GETTERS AND SETTERS ************************************************
    //**********************************************************************
    public String getAnalytical_rep_id() {
        return analytical_rep_id;
    }

    public void setAnalyticalReplicateID(String bioreplicate_id, int nAnalyticalRep) {
        this.bioreplicate_id = bioreplicate_id;
        this.analytical_rep_id = "AR" + bioreplicate_id.substring(2) + "." + String.format("%03d", nAnalyticalRep);
    }

    public void setAnalyticalReplicateID(String analytical_rep_id) {
        this.analytical_rep_id = analytical_rep_id;
    }

    public String getAnalyticalReplicateName() {
        return analytical_rep_name;
    }

    public void setAnalyticalReplicateName(String analytical_rep_name) {
        this.analytical_rep_name = analytical_rep_name;
    }

    public String getBioreplicateID() {
        return bioreplicate_id;
    }

    public void setBioreplicateID(String bioreplicate_id) {
        this.bioreplicate_id = bioreplicate_id;
    }

    public String getTreatmentID() {
        return treatment_id;
    }

    public void setTreatmentID(String treatment_id) {
        this.treatment_id = treatment_id;
    }

    public String getStatus() {
        return status;
    }

    public void setStatus(String status) {
        this.status = status;
    }

    //***********************************************************************
    //* OTHER FUNCTIONS *****************************************************
    //***********************************************************************
    @Override
    public String toString() {
        return this.toJSON();
    }
}

File: web_old/app/model/SampleModels/AnalyticalReplicate.js
/*
 * (C) Copyright 2014 The Genomics of Gene Expression Lab, CIPF 
 * (http://bioinfo.cipf.es/aconesawp) and others.
 *
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the GNU Lesser General Public License
 * (LGPL) version 3 which accompanies this distribution, and is available at
 * http://www.gnu.org/licenses/lgpl.html
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * Contributors:
 *     Rafael Hernandez de Diego, rhernandez@cipf.es
 *     Ana Conesa Cegarra, aconesa@cipf.es
 *     
 * THIS FILE CONTAINS THE FOLLOWING COMPONENT DECLARATION
 * - AnalyticalReplicate
 */
Ext.define('SL.model.SampleModels.AnalyticalReplicate', {
    extend: 'Ext.data.Model',
    alias: "widget.AnalyticalReplicate",
    //Extends the Observable class
    mixins: {Observable: 'SL.view.senchaExtensions.Observable'},
    fields: [
        {name: 'analytical_rep_id'},
        {name: 'analytical_rep_name'},
        {name: 'bioreplicate_id'},
        {name: 'biocondition_id'},
        {name: 'treatment_id'}
    ],
    getID: function () {
        return this.get("analytical_rep_id");
    },
    setID: function (analyticalRepID) {
        this.set("analytical_rep_id", analyticalRepID);
    },
    getAnalyticalRepID: function () {
        return this.get("analytical_rep_id");
    },
    setAnalyticalRepID: function (analyticalRepID) {
        this.set("analytical_rep_id", analyticalRepID);
    },
    getBioReplicateID: function () {
        return this.get("bioreplicate_id");
    },
    setBioReplicateID: function (analyticalRepName) {
        this.set("bioreplicate_id", analyticalRepName);
    },
    getAnalyticalRepName: function () {
        return this.get("analytical_rep_name");
    },
    setAnalyticalRepName: function (analyticalRepName) {
        this.set("analytical_rep_name", analyticalRepName);
    },
    getTreatmentID: function () {
        return this.get("treatment_id");
    },
    setTreatmentID: function (treatmentID) {
        this.set("treatment_id", treatmentID);
    },
    toSimpleJSON: function () {
        var JSON_DATA = this.data;
        return JSON_DATA;
    },
    loadFromJSON: function (jsonData, ignoreIDs) {
        if (ignoreIDs != null) {
            if (jsonData['analytical_rep_id']) {
                delete jsonData['analytical_rep_id'];
            }
        }
        this.set(jsonData);
        return this;
    },
    getJSONforGraph: function () {
        var nodeData = {
            id: this.getAnalyticalRepID(),
            label: (this.getAnalyticalRepName() != null ? this.getAnalyticalRepName().substring(0, 20) + (this.getAnalyticalRepName().length > 21 ? "..." : "") : ""),
            type: 'analyticalsample',
            description: "<b>Analytical Sample</b></br><i>" + (this.getAnalyticalRepName() != null ? this.getAnalyticalRepName() : "") + "</i>",
            parent: this.getBioReplicateID() + "-" + this.getTreatmentID()
        };

        return nodeData;
    },
    clone: function () {
        var newInstance = this.copy();
        var observers = this.getObservers();
        for (var i in observers) {
            newInstance.addObserver(observers[i]);
        }

        return newInstance;
    },
    getMemento: function () {
        return this.clone();
    }
});